File: modules/filtering/js/ts/filtering.ts
/*
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/**
 * Anisotropic smoothing of mesh geometry and per-vertex scalar fields.
 * Uses a normal/vertex diffusion formulation that preserves sharp features
 * better than naive Laplacian smoothing.
 *
 * Omitted option fields fall back to library defaults.
 */

import type { SurfaceMesh } from "./core.js";

/**
 * - `"normalSmoothing"`: diffuse facet normals, then fit geometry to them.
 *   Preserves sharp features better.
 * - `"vertexSmoothing"`: diffuse vertex positions directly. Faster, more aggressive.
 */
export type SmoothingMethod = "normalSmoothing" | "vertexSmoothing";

export interface MeshSmoothingOptions {
  /** Smoothing strategy. Default: `"normalSmoothing"`. */
  method?: SmoothingMethod;
  /**
   * Curvature-based feature weight in `[0, 1]`. Higher values protect
   * high-curvature regions (edges, corners) from being flattened.
   * Default: `0.02`.
   */
  curvatureWeight?: number;
  /** Diffusion time-step for normals; larger = more smoothing per call. Default: `0.1`. */
  normalSmoothingWeight?: number;
  /** Trade-off between matching smoothed normals and keeping positions. Default: `1`. */
  gradientWeight?: number;
  /** Gradient scaling: `<1` smooths further, `>1` sharpens. Default: `0`. */
  gradientModulationScale?: number;
  /** How strongly geometry is pulled toward the target normal field. Default: `0.1`. */
  normalProjectionWeight?: number;
}

export interface AttributeSmoothingOptions {
  /** Curvature-based feature weight in `[0, 1]`; higher = protect features more. Default: `0.02`. */
  curvatureWeight?: number;
  /** Diffusion time-step for normals used to build the anisotropic metric. Default: `0.1`. */
  normalSmoothingWeight?: number;
  /** Trade-off between matching gradients and keeping values. Default: `1`. */
  gradientWeight?: number;
  /** Gradient scaling; `<1` smooths more, `>1` sharpens. Default: `0`. */
  gradientModulationScale?: number;
}

/**
 * Smoothing / denoising. Accessible as `lagrange.filtering`.
 */
export interface FilteringModule {
  /**
   * Feature-preserving mesh smoothing. Moves vertex positions so the surface
   * becomes smoother while preserving corners and creases. In-place.
   */
  meshSmoothing(mesh: SurfaceMesh, opts?: MeshSmoothingOptions): void;
  /**
   * Smooth a named per-vertex scalar attribute across the surface. Useful for
   * cleaning up painted weights, noisy curvature, etc. In-place.
   */
  scalarAttributeSmoothing(mesh: SurfaceMesh, attributeName: string, opts?: AttributeSmoothingOptions): void;
}

export const filteringModuleKeys = [
  "meshSmoothing",
  "scalarAttributeSmoothing",
] as const satisfies readonly (keyof FilteringModule)[];

File: modules/filtering/js/src/filtering.cpp
/*
 * Copyright 2026 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

#include "bind_types.h"

#include <lagrange/filtering/attribute_smoothing.h>
#include <lagrange/filtering/mesh_smoothing.h>

#include <emscripten/bind.h>
#include <emscripten/val.h>

#include <string>

namespace {

using namespace lagrange;
using namespace lagrange::js::bind;
using val = emscripten::val;

} // namespace

EMSCRIPTEN_BINDINGS(lagrange_filtering)
{
    using namespace emscripten;

    function(
        "meshSmoothing",
        +[](MeshType& mesh, val opts) {
            filtering::SmoothingOptions o;
            if (!opts.isUndefined()) {
                auto method = opts["method"];
                if (!method.isUndefined()) {
                    auto s = method.as<std::string>();
                    if (s == "vertexSmoothing") {
                        o.filter_method =
                            filtering::SmoothingOptions::FilterMethod::VertexSmoothing;
                    } else {
                        o.filter_method =
                            filtering::SmoothingOptions::FilterMethod::NormalSmoothing;
                    }
                }
                apply_opt(opts, "curvatureWeight", o.curvature_weight);
                apply_opt(opts, "normalSmoothingWeight", o.normal_smoothing_weight);
                apply_opt(opts, "gradientWeight", o.gradient_weight);
                apply_opt(opts, "gradientModulationScale", o.gradient_modulation_scale);
                apply_opt(opts, "normalProjectionWeight", o.normal_projection_weight);
            }
            filtering::mesh_smoothing(mesh, o);
        });

    function(
        "scalarAttributeSmoothing",
        +[](MeshType& mesh, const std::string& attribute_name, val opts) {
            filtering::AttributeSmoothingOptions o;
            if (!opts.isUndefined()) {
                apply_opt(opts, "curvatureWeight", o.curvature_weight);
                apply_opt(opts, "normalSmoothingWeight", o.normal_smoothing_weight);
                apply_opt(opts, "gradientWeight", o.gradient_weight);
                apply_opt(opts, "gradientModulationScale", o.gradient_modulation_scale);
            }
            filtering::scalar_attribute_smoothing(mesh, attribute_name, o);
        });
}
